File: DesignPrinciples_Session2/4_SOLID/DIP/CSharp/Bad_OrderService.cs
// DIP — VIOLAZIONE: la logica di business dipende direttamente da dettagli tecnologici.
// Per cambiare DB, SMTP o Stripe → toccare OrderService.
// Per testare OrderService → serve un DB reale, un server SMTP, un account Stripe.

public class OrderService
{
    // Dipendenze hardcoded: costruite internamente, non iniettabili
    private readonly SqlOrderRepository _repository = new SqlOrderRepository();
    private readonly SmtpEmailSender _emailSender = new SmtpEmailSender();
    private readonly StripePaymentGateway _payments = new StripePaymentGateway();

    public void Place(Order order)
    {
        if (!order.Items.Any())
            throw new InvalidOperationException("Ordine vuoto");

        var paymentResult = _payments.Charge(order.Total, order.Customer.Card);
        if (!paymentResult.Success)
            throw new PaymentFailedException(paymentResult.Error);

        _repository.Save(order);
        _emailSender.Send(order.Customer.Email, "Conferma ordine");
    }
}

// Classi concrete (dettagli tecnologici)
public class SqlOrderRepository
{
    public void Save(Order order) { /* SQL Server */ }
}

public class SmtpEmailSender
{
    public void Send(string to, string subject) { /* SMTP */ }
}

public class StripePaymentGateway
{
    public PaymentResult Charge(decimal amount, object card) => new PaymentResult(true, null);
}

File: DesignPrinciples_Session2/4_SOLID/DIP/Python/bad_order_service.py
# DIP — VIOLAZIONE: OrderService crea direttamente le dipendenze concrete.
# Non testabile senza DB, SMTP e Stripe reali.

import sqlite3
import smtplib
import stripe


class OrderService:
    def place(self, order):
        if not order.items:
            raise ValueError("Ordine vuoto")

        # Dipendenza diretta da Stripe
        charge = stripe.Charge.create(
            amount=int(order.total * 100),
            currency="eur",
            source=order.customer.card_token,
        )
        if charge.status != "succeeded":
            raise RuntimeError("Pagamento fallito")

        # Dipendenza diretta da SQLite
        conn = sqlite3.connect("orders.db")
        conn.execute("INSERT INTO orders ...", (order.id,))
        conn.commit()

        # Dipendenza diretta da SMTP
        with smtplib.SMTP("smtp.azienda.it") as smtp:
            smtp.sendmail("noreply@azienda.it", order.customer.email, "Conferma ordine")
